pkg/migration-tool/planner/planner.go

```diff
@@ -0,0 +1,238 @@
+	}
+	readClient, err := utils.NewClient("reader-last-maxt-pushed", c.ProgressMetricURL, utils.Write, model.Duration(time.Minute*2))
+	if err != nil {
+		return -1, false, fmt.Errorf("create fetch-last-pushed-maxt reader: %w", err)
+	}
+	result, _, _, err := readClient.Read(context.Background(), query)
+	if err != nil {
+		return -1, false, fmt.Errorf("fetch-last-pushed-maxt query result: %w", err)
+	}
+	ts := result.Timeseries
+	if len(ts) == 0 {
+		return -1, false, nil
+	}
+	for _, series := range ts {
+		for i := len(series.Samples) - 1; i >= 0; i-- {
+			if series.Samples[i].Timestamp > lastPushedMaxt {
+				lastPushedMaxt = series.Samples[i].Timestamp
+			}
+		}
+	}
+	if lastPushedMaxt == 0 {
+		return -1, false, nil
+	}
+	return lastPushedMaxt, true, nil
+}
+
+func (p *Plan) DecrementBlockCount() {
+	p.blockCounts.Sub(1)
+}
+
+// ShouldProceed reports whether the fetching process should proceeds further. If any time-range is left to be
+// fetched from the provided time-range, it returns true, else false.
+func (p *Plan) ShouldProceed() bool {
+	return p.nextMint < p.config.Maxt
+}
+
+// update updates the details of the planner that are dependent on previous fetch stats.
+func (p *Plan) update(numBytes int) {
+	p.lastNumBytes.Store(int64(numBytes))
+}
+
+func (p *Plan) LastMemoryFootprint() int64 {
+	return p.lastNumBytes.Load()
+}
+
+// NewBlock returns a new block after allocating the time-range for fetch.
+func (p *Plan) NextBlock() (reference *Block, err error) {
+	timeDelta := determineTimeDelta(p.lastNumBytes.Load(), p.config.BlockSizeLimitBytes, p.lastTimeRangeDelta)
+	mint := p.nextMint
+	maxt := mint + timeDelta
+	if maxt > p.config.Maxt {
+		maxt = p.config.Maxt
+	}
+	p.nextMint = maxt
+	p.lastTimeRangeDelta = timeDelta
+	bRef, err := p.createBlock(mint, maxt)
+	if err != nil {
+		return nil, fmt.Errorf("next-block: %w", err)
+	}
+	return bRef, nil
+}
+
+// createBlock creates a new block and returns reference to the block for faster write and read operations.
+func (p *Plan) createBlock(mint, maxt int64) (reference *Block, err error) {
+	if err = p.validateT(mint, maxt); err != nil {
+		return nil, fmt.Errorf("create-block: %w", err)
+	}
+	id := p.blockCounts.Add(1)
+	timeRangeInMinutes := (maxt - mint) / time.Minute.Milliseconds()
+	percent := float64(maxt-p.config.Mint) * 100 / float64(p.config.Maxt-p.config.Mint)
+	if percent > 100 {
+		percent = 100
+	}
+	baseDescription := fmt.Sprintf("progress: %.3f%% | block-%d time-range: %d mins | mint: %d | maxt: %d", percent, id, timeRangeInMinutes, mint, maxt)
+	reference = &Block{
+		id:                    id,
+		percent:               percent,
+		pbarDescriptionPrefix: baseDescription,
+		pbar: progressbar.NewOptions(
+			6,
+			progressbar.OptionOnCompletion(func() {
+				_, _ = fmt.Fprint(os.Stderr, "\n")
+			}),
+		),
+		mint:    mint,
+		maxt:    maxt,
+		pbarMux: p.pbarMux,
+		plan:    p,
+	}
+	if p.Quite {
```

Comment:
Misspelled variable name
```suggestion
	if p.Quiet {
```